enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,480 @@
+  
+- Implement eBPF tabels loader and attach to the selected interfaces.
+
+- Implement implicit firewall pinholes rules to allow fail safe in case of invalid configuration (user shoudln't try to overwrite those rules).
+
+- Implement verification webhook to validate configuration and protect against invalid rules.
+  
+- Create Ingress node eBPF OCP repro.
+
+- Implement and build Ingress node eBPF kernel hook to do ingress node firewall.
+
+- Ingress node firewall E2E testing.
+
+### API Extensions
+
+A new cluster scoped CRD is introduced to support this feature this CRD will be managed by `ClusterNetworkOperator`
+```golang
+type IngressNodeFirewallFromCIDR struct {
+	// FromCIDR is the IPv4/IPv6 CIDR from which we apply node firewall rule
+	// +kubebuilder:validation:one-of[]=Type=string;Format:ipv4;Format:ipv6
+	FromCIDR string `json:"fromCIDR" protobuf:"bytes,1,opt,name=fromCIDR"`
+}
+
+// IngressNodeFirewallICMPRule define ingress node firewall rule for ICMP and ICMPv6 protocols
+type IngressNodeFirewallICMPRule struct {
+	// ICMPType define ICMP Type Numbers (RFC 792).
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Maximum:=43
+	// +kubebuilder:validation:Minimum:=0
+	ICMPType int8 `json:"icmpType" protobuf:"varint,1,opt,name=icmpType"`
+
+	// ICMPCode define ICMP Code ID (RFC 792).
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Maximum:=16
+	// +kubebuilder:validation:Minimum:=0
+	ICMPCode int8 `json:"icmpCode" protobuf:"varint,2,opt,name=icmpCode"`
+}
+
+// IngressNodeFirewallProtoRule define ingress node firewall rule for TCP, UDP and SCTP protocols
+type IngressNodeFirewallProtoRule struct {
+    // start-end range of dstPorts or [port1, port2,..., portn].
+    // +kubebuilder:validation:Optional
+    // +optional
+    Ports []int32 `json:"ports" protobuf:"varint,2,opt,name=ports"`
+}
+
+// IngressNodeFirewallProtocolRule define ingress node firewall rule per protocol
+type IngressNodeFirewallProtocolRule struct {
+    // IngressNodeFirewallProtoRule define ingress node firewall rule for TCP, UDP and SCTP protocols.
+	ProtocolRule IngressNodeFirewallProtoRule json:"protoRule" protobuf:"varint,1,opt,name=protoRule"`
+    
+	// IngressNodeFirewallICMPRule define ingress node firewall rule for ICMP and ICMPv6 protocols.
+    ICMPRule IngressNodeFirewallICMPRule
+
+	// Protocol can be ICMP, ICMPv6, TCP, SCTP or UDP.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="ICMP";"ICMPv6";"TCP";"UDP";"SCTP"
+	Protocol IngressNodeFirewallRuleProtocolType
+
+	// Action can be Allow or Deny.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="Allow";"Deny"
+	Action IngressNodeFirewallActionType `json:"action" protobuf:"bytes,4,opt,name=action,casttype=IngressNodeFirewallActionType"`
+}
+
+// ProtocolType defines the protocol types that are supported
+type IngressNodeFirewallRuleProtocolType string
```

Comment:
```suggestion
// IngressNodeFirewallRuleProtocolType defines the protocol types that are supported
type IngressNodeFirewallRuleProtocolType string
```
Looks like you need to change `ProtocolType` to `IngressNodeFirewallRuleProtocolType` below too (or change `IngressNodeFirewallRuleProtocolType` to `ProtocolType` here).

Reply:
yeah fixed that during PoC